tests/jni_weak_refs.rs

```diff
@@ -0,0 +1,84 @@
+#![cfg(feature = "invocation")]
+
+use std::{
+    sync::{Arc, Barrier},
+    thread::spawn,
+};
+
+use jni::{
+    objects::{AutoLocal, JValue},
+    sys::jint,
+};
+
+mod util;
+use util::{attach_current_thread, unwrap};
+
+#[test]
+pub fn global_ref_works_in_other_threads() {
```

Comment:
I'd suggest a couple of simpler tests to verify the basic functionality — this one is rather complex:

1. Create a WeakGlobalRef (and a GlobalRef, to keep the object around), detach the current thread, re-attach it, and verify it can be upgraded to strong local and global references.
2. Create a WeakGlobalRef from `JObject::null`.
  - :warning: A design question: shall JNIEnv create a WeakGlobalRef that can never be upgraded, or return Option::none?
3. (A more advanced test, similar to the ones we use for GlobalRefs) Create a VM with small memory limit (e.g., 32MB), create a WeakGlobalRef to a large-ish object (e.g., 16MB), release any strong refs to it, create a different large-ish object, so that the first is GCed (e.g., 16MB will certainly do the job), verify the WeakGlobalRef cannot be upgraded.
4. (May keep this one) Test it is accessible from different threads.